packages/next/src/server/app-render/app-render-prerender-utils.ts

```diff
@@ -0,0 +1,226 @@
+ * in the edge builds since this might also avoid setTimeout throttling.
+ */
+export function waitAtLeastOneReactRenderTask(): Promise<void> {
+  if (process.env.NEXT_RUNTIME === 'edge') {
+    return new Promise((r) => setTimeout(r, 0))
+  } else {
+    return new Promise((r) => setImmediate(r))
+  }
+}
+
+/**
+ * This is a utility function to make scheduling sequential tasks that run back to back easier.
+ * We schedule on the same queue (setImmediate) at the same time to ensure no other events can
+ * sneak in between. We pass the return value from the first one to the second one to make typing
+ * easier.
+ *
+ * We do not call the second function if the first errored. practiaclly spea
+ */
+export function prerenderAndAbortInSequentialTasks<R>(
+  prerender: () => Promise<R>,
+  abort: () => void
+): Promise<R> {
+  if (process.env.NEXT_RUNTIME === 'edge') {
+    throw new Error(
+      'schedulePrerendreWork should not be called in edge runtime. This is a bug in Next.js'
+    )
+  }
+  return new Promise((resolve, reject) => {
+    let pendingResult: Promise<R>
+    setImmediate(() => {
+      try {
+        pendingResult = prerender()
+      } catch (err) {
+        reject(err)
+      }
+    })
+    setImmediate(() => {
+      abort()
+      resolve(pendingResult)
+    })
+  })
+}
+
+// React's RSC prerender function will emit an incomplete flight stream when using `prerender`. If the connection
+// closes then whatever hanging chunks exist will be errored. This is because prerender (an experimental feature)
+// has not yet implemented a concept of resume. For now we will simulate a paused connection by wrapping the stream
+// in one that doesn't close even when the underlying is complete.
+export class ReactServerResult {
+  private _stream: null | ReadableStream<Uint8Array>
+
+  constructor(stream: ReadableStream<Uint8Array>) {
+    this._stream = stream
+  }
+
+  tee() {
+    if (this._stream === null) {
+      throw new Error(
+        'Cannot tee a ReactServerResult that has already been consumed'
+      )
+    }
+    const tee = this._stream.tee()
+    this._stream = tee[0]
+    return tee[1]
+  }
+
+  consume() {
+    if (this._stream === null) {
+      throw new Error(
+        'Cannot consume a ReactServerResult that has already been consumed'
+      )
+    }
+    const stream = this._stream
+    this._stream = null
+    return stream
+  }
+}
+
+type ReactPromise<T> = Promise<T> & {
+  status?: string
+  value?: T
+  reason?: unknown
+}
+export type ReactServerPrerenderResolveToType = {
+  prelude: ReadableStream<Uint8Array>
+}
+
+export async function createReactServerPrerenderResult(
+  underlying: Promise<ReactServerPrerenderResolveToType>
+): Promise<ReactServerPrerenderResult>
+export async function createReactServerPrerenderResult(
+  underlying: ReactPromise<ReactServerPrerenderResolveToType>
+): Promise<ReactServerPrerenderResult> {
```

Comment:
AFAICT the `ReactPromise` overload is redundant, there seems to be no `ReactPromise`-specific handling inside here, and it's never called with a `ReactPromise`. so i think we can just drop it?
```suggestion
): Promise<ReactServerPrerenderResult> {
```